feature: TrailingCommaInMultilineFixer - Add array destructuring support

SpacePossum opened the pull request.

julienfalque commented on doc/rules/basic/trailing_comma_in_multiline.rst:
```diff
@@ -124,7 +124,11 @@ The rule is part of the following rule sets:
-  Using the `@Symfony <./../../ruleSets/Symfony.rst>`_ rule set will enable the ``trailing_comma_in_multiline`` rule with the default config.
+  Using the `@Symfony <./../../ruleSets/Symfony.rst>`_ rule set will enable the ``trailing_comma_in_multiline`` rule with the config below:
+
+  ``['elements' => ['array_destructuring', 'arrays', 'match']]``
```
The default value (as was previously used) is `['arrays']`.

SpacePossum replied:
This is a new rule, as such there is no "previously used"?

julienfalque replied:
> This is a new rule

Not sure I follow you. What makes you think the rule is new?

Anyway, the `@Symfony` ruleset isn't new so it shouldn't change. But actually Symfony might be OK with this? @nicolas-grekas What do you think?

keradus replied:
`trailing_comma_in_multiline` rule is already on master, so it is not a new rule indeed

yet, the configuration of it for Sf ruleset changed - not sure if it's welcome by Sf community, did we checked if repo follows this new rule already, or checked with maintainers?

SpacePossum replied:
I confused this PR with another, sorry about that.
The changes on SF 5.4 would look like this:
https://github.com/SpacePossum/symfony/pull/2/files

julienfalque commented on doc/rules/basic/trailing_comma_in_multiline.rst:
```diff
@@ -124,7 +124,11 @@ The rule is part of the following rule sets:
-  Using the `@PhpCsFixer <./../../ruleSets/PhpCsFixer.rst>`_ rule set will enable the ``trailing_comma_in_multiline`` rule with the default config.
+  Using the `@PhpCsFixer <./../../ruleSets/PhpCsFixer.rst>`_ rule set will enable the ``trailing_comma_in_multiline`` rule with the config below:
+
+  ``['elements' => ['array_destructuring', 'arrays', 'match']]``
```
We can add `'array_destructuring'` here.

julienfalque commented on src/Fixer/Basic/TrailingCommaInMultilineFixer.php:
```diff
@@ -12,15 +12,14 @@
  */
 
-namespace PhpCsFixer\Fixer\ControlStructure;
+namespace PhpCsFixer\Fixer\Basic;
```
The class is not marked as internal, isn't this a BC break?

SpacePossum replied:
Do we hold a BC promise on any implementation of a single non-abstract rule?

julienfalque replied:
TBH I don't know 😅️

@keradus WDYT?

keradus replied:
I would name it BC break, following the rules I was following so far (no `@internal`, then under BC promise).

SpacePossum replied:
I move it back. The reason was that this fixer has nothing to do with `ControlStructure` IMHO.

SpacePossum commented:
rebased on latest, fixed conflict, moved back so stuff (https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/6518#discussion_r934743037), no other changes

jrmajor commented on doc/rules/control_structure/trailing_comma_in_multiline.rst:
```diff
@@ -23,7 +23,7 @@ Default value: ``false``
 ``match``).
 
-Allowed values: a subset of ``['arrays', 'arguments', 'parameters', 'match']``
+Allowed values: a subset of ``['arguments', 'arrays', 'array_destructuring', 'match', 'parameters']``
```
Is there a need for a separate option? I think it would make sense to fix array destructuring when the `arrays` option is enabled.

SpacePossum replied:
It is mainly for BC reasons, now the new feature of fixing array destructing is opt-in. We can always consider to merge two functionalities under one config option in the next major release.

jrmajor replied:
We have BC breaks like this every release, e.g. we've just merged #6583. BTW, do we have a place to keep track of things we want to consider before the next major?

SpacePossum replied:
>do we have a place to keep track of things we want to consider before the next major?

not that I know of

> We have BC breaks like this every release, e.g. we've just merged https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/6583.

I don't think there is a BC break in that PR

On the topic of this PR, what is the problem with the current proposal?

jrmajor replied:
> I don't think there is a BC break in that PR

Neither do I, some fixer just learned a new trick, just like in this one :) I fail to see how they're different.

Both for users and for us it's easier not to include an option if it's unnecessary than to remove it in the next major.

I do not want to block this PR, if you feel that the BC break would be too significant feel free to ignore my comments :)

SpacePossum commented:
rebased on latest, no other changes

SpacePossum commented:
rebased on latest, no other changes
